File: src/functions/generate_question.py
def generate_question(level):
    conn_str = [" ∨ ", " ∧ ", " → ", " ↔ "]

    def conFunc(P, Q):
        return (P and Q)

    def disFunc(P, Q):
        return (P or Q)

    def implyFunc(P, Q):
        return ((not P) or Q)

    def biimplyFunc(P, Q):
        return (P == Q)

    conn_funcs = [disFunc, conFunc, implyFunc, biimplyFunc]

    n_prop = level + 1
    V = []
    for i in range(n_prop):
        V.append(random.choice([chr(random.randint(65, 65 + i)), chr(65 + i)]))
    V_s = V.copy()

    prop_val_dict = {}

    V_no_dup = list(set(V))

    for var in V_no_dup:
        prop_val_dict[var] = random.randint(0, 1)

    question_elements = []
    elements_answers = []

    for index, value in enumerate(V):
        V[index] = prop_val_dict[value]

    while len(V) > 1:
        i = random.randint(0, len(V) - 2)
        c_i = random.randint(0, len(conn_str) - 1)

        V_s[i] = "(" + V_s[i] + conn_str[c_i] + V_s[i + 1] + ")"
        del V_s[i + 1]

        question_elements.append(V_s[i][1:-1])

        V[i] = conn_funcs[c_i](V[i], V[i + 1])
        del V[i + 1]

        elements_answers.append(V[i])

    print(prop_val_dict)
    print(question_elements)
    print(elements_answers)
    return prop_val_dict, question_elements, elements_answers

File: src/functions/generateQuestion.js
const generateQuestion = (level) => {
  const connStr = [" ∨ ", " ∧ ", " → ", " ↔ "];

  function conFunc(P, Q) {
    return P && Q;
  }

  function disFunc(P, Q) {
    return P || Q;
  }

  function implyFunc(P, Q) {
    return !P || Q;
  }

  function biimplyFunc(P, Q) {
    return P === Q;
  }

  const connFuncs = [disFunc, conFunc, implyFunc, biimplyFunc];

  const nProp = level + 1;
  const V = [];
  for (let i = 0; i < nProp; i++) {
    V.push(String.fromCharCode(Math.floor(Math.random() * (65 + i - 65 + 1)) + 65));
  }
  const V_s = [...V];

  const propValDict = {};

  const VNoDup = [...new Set(V)];

  VNoDup.forEach((varName) => {
    propValDict[varName] = Math.round(Math.random());
  });

  const questionElements = [];
  const elementsAnswers = [];

  V.forEach((value, index) => {
    V[index] = propValDict[value];
  });

  while (V.length > 1) {
    const i = Math.floor(Math.random() * (V.length - 1));
    const c_i = Math.floor(Math.random() * connStr.length);

    V_s[i] = `(${V_s[i]}${connStr[c_i]}${V_s[i + 1]})`;
    V_s.splice(i + 1, 1);

    questionElements.push(V_s[i].substring(1, V_s[i].length - 1));

    V[i] = connFuncs[c_i](V[i], V[i + 1]);
    V.splice(i + 1, 1);

    elementsAnswers.push(V[i]);
  }

  return [propValDict, questionElements, elementsAnswers];
}

export default generateQuestion
